### ISSUE TITLE
Rule proposal: Ban implicit conversion of Readonly<T> to T

### ISSUE THREAD

**Issue by bkietz** [typescript-eslint/typescript-eslint#11833]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [x] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

TypeScript silently allows an expression of type `Readonly<T>` to be passed as an argument of type `T`. This results in a loss of the readonly type information in the scope of the called function and potential unintended mutation.

### Fail Cases

```TypeScript
// Declaring a variable Readonly should guard against future modification, even at a lower scope
const dangerous: Readonly<number[]> = getDangerousToModify();

// Fails since pushOne() requires a mutable argument
pushOne(dangerous);
function pushOne(a: number[]): void { a.push(1); };
```

### Pass Cases

```TypeScript
// Making a copy results in a mutable array, so the rule accepts this
const dangerousCopy = [...dangerous];
pushOne(dangerousCopy);

// If mutation is really intended, type assertion will also be accepted
pushOne(dangerous as number[]);
```

### Additional Info

In combination with [explicit `this` parameters](https://www.typescriptlang.org/docs/handbook/2/functions.html#declaring-this-in-a-function), this also enables [methods](https://www.typescriptlang.org/play/?#code/MYGwhgzhAEBqCmwDM0DeAoa1gHsB2EALgE4CuwhOxAFAA6kBGIAlsNAB4Bc0epAtg3jEANNHpNW0AJ7deAoaPEs2AL1n9BxAJRoAvukw8qfaoQAWzCNwBK8MABN8IKQB5zlgHxb184mkNYxPCEpMR40ACyYOYAdGZStDiEphYQMeyi7mlSmakxKloA3Ib6hnjGYCwq8CmW3AjI3tAAbjjM9v5YWLgEhDzQALzQWTHlxCZFAcN57NAA9EN4xV3TljFS84vLXSMqmzzb+qUAZqR4FMz4RuOVzNUAcvgR8L7UzfWISE2t7Z3Y+ER+kNmqNjNRJlgQbMFgdDCCNjClnD8vskaU5nNoKdzoRLuFcMQghRnABJEAgeAAc0qACFSIQADLMPjMQjwezUYg4Gx2Rx4ZwuBpILzcH4dDAYro9QHhIZc0HjcHFSWBHDpVHKzGq9Ya9Aq6DyvaIzXQE5nC5XCBmHCkED2GnwMkU6kgAAMnO50FsDicriFIpabXFZQqVXgjzwz1eXKK80xtEgEGYeEpXt5vsFnw80EgcE+hWg9kswDAxHsMGlfWTxwquPw6DNOLx0CtNrtDqdVMqAEYPTyffy-VnvkG-vKxnxbtUlXGxInk6nvXyBf6czAhQWixAS2WKwCq3gazc63gGwYgA) to declare they are guaranteed not to modify the object:

```ts
class Vec3 {
  constructor(public x: number, public y: number, public z: number) {}

  norm(this: Readonly<this>): number {
    return Math.hypot(this.x, this.y, this.z);
  }
}
```

**Comment by jendrikw** [comment]
Previous discussions: #11422 #5954

**Comment by bkietz** [comment]
Closing as a duplicate
